src/corecel/sys/KernelDiagnostics.cc

```diff
@@ -17,6 +19,18 @@
 
 namespace celeritas
 {
+//---------------------------------------------------------------------------//
+/*!
+ * Try adding a new kernel in a thread-safe fashion.
+ */
+auto KernelDiagnostics::threadsafe_insert_new(std::uintptr_t fptr)
+    -> InsertResult
+{
+    static std::mutex           insert_mutex;
+    std::lock_guard<std::mutex> scoped_lock{insert_mutex};
+    return keys_.insert({fptr, key_type{this->size()}});
```

Comment:
The same mutex should be used to protect any read access to `keys_` otherwise the underlying data structure might be being modified at the same time it is being read.

Reply:
Ugh, though that means I should be adding the same mutex to all the `values_` at launch time as well... 🤔 yuck.

Reply:
on the other hand, locking all the time "just" for an ensure is pretty bad ...

Reply:
The problem isn't the ensure, it's that whenever we launch a kernel we access `values_` by kernel ID. If one thread happens to be adding a new kernel (and say it has to allocate the under-the-hood storage) while another is incrementing an existing kernel, the code will crash. And we don't want to mutex every kernel launch twice, especially since the writing only happens during the first step. An alternative would be to make a vector of shared pointers, and replace `KernelId` with `shared_ptr<Diagnostics>`...

Reply:
`values_` indeed has in itself the same problem as `keys_` and the assignment to the value in the map and the `push_back` inside `values_` must absolutely be done `at the same time`. eg
instead of:
```
values_.push_back(std::move(diag));
```
we could use under the mutex:
```
auto keyloc = iter_inserted.first.second.get()
values_.resize(keyloc + 1)
values_[keyloc] = std::move(diag));
```

We have to remove:
```
CELER_ENSURE(keys_.size() == values_.size());
```
which is both not that useful in single thread (it basically test that `unordered_map` return value is "correct", otherwise there is no way not get to the push_back). (I suspect that a better place for it would be *before* the insert in the map, i.e. testing that some other piece of code did not get the 2 containers out of sync)

Reply:
> An alternative would be to make a vector of shared pointers

I don't see how that helps ... (but maybe my previous comment is answering)